File: packages/sandbox/daemon/git/porcelain.ts
/** Parse one `git status --porcelain=v1 -z` entry into status chars + path. */
export function parsePorcelainEntry(
  entry: string,
): { index: string; working: string; path: string } | null {
  if (entry.length < 3) return null;
  const index = entry[0] ?? " ";
  const working = entry[1] ?? " ";
  // Standard v1: "XY filename". Some `-z` outputs omit the separator space.
  const path =
    entry.length >= 4 && entry[2] === " " ? entry.slice(3) : entry.slice(2);
  if (!path) return null;
  return { index, working, path };
}

/** Parse full `-z` porcelain output into a set of changed paths. */
export function parsePorcelainZ(out: string): Set<string> {
  const paths = new Set<string>();
  const parts = out.split("\0");
  for (let i = 0; i < parts.length; i++) {
    const entry = parts[i];
    if (!entry) continue;
    const parsed = parsePorcelainEntry(entry);
    if (!parsed) continue;
    paths.add(parsed.path);
    if (parsed.index === "R" || parsed.index === "C") {
      i++;
      const orig = parts[i];
      if (orig) paths.add(orig);
    }
  }
  return paths;
}

File: packages/sandbox/daemon-go/internal/gitx/porcelain.go
package gitx

import "strings"

type PorcelainFile struct {
	Path       string `json:"path"`
	Index      string `json:"index"`
	WorkingDir string `json:"working_dir"`
}

func parsePorcelainEntry(entry string) (PorcelainFile, bool) {
	if len(entry) < 3 {
		return PorcelainFile{}, false
	}
	index := string(entry[0])
	working := string(entry[1])
	var path string
	if len(entry) >= 4 && entry[2] == ' ' {
		path = entry[3:]
	} else {
		path = entry[2:]
	}
	if path == "" {
		return PorcelainFile{}, false
	}
	return PorcelainFile{Path: path, Index: index, WorkingDir: working}, true
}

func ParsePorcelainFiles(out string) []PorcelainFile {
	files := []PorcelainFile{}
	parts := strings.Split(out, "\x00")
	for i := 0; i < len(parts); i++ {
		entry := parts[i]
		if entry == "" {
			continue
		}
		parsed, ok := parsePorcelainEntry(entry)
		if !ok {
			continue
		}
		files = append(files, parsed)
		if parsed.Index == "R" || parsed.Index == "C" {
			i++
		}
	}
	return files
}

func ParsePorcelainZ(out string) map[string]struct{} {
	paths := map[string]struct{}{}
	parts := strings.Split(out, "\x00")
	for i := 0; i < len(parts); i++ {
		entry := parts[i]
		if entry == "" {
			continue
		}
		parsed, ok := parsePorcelainEntry(entry)
		if !ok {
			continue
		}
		paths[parsed.Path] = struct{}{}
		if parsed.Index == "R" || parsed.Index == "C" {
			i++
			if i < len(parts) && parts[i] != "" {
				paths[parts[i]] = struct{}{}
			}
		}
	}
	return paths
}
